### PR TITLE
Unhiding a tag should unfollow it too

### LINKED ISSUE #20039: Unhiding a tag should unfollow it too
Related to [Hide tag weights and change negative weight tags to "Hidden tags" #19675
](https://github.com/forem/forem/issues/19675)

## Is this feature related to a problem? Please describe.

In the [Hide tag weights and change negative weight tags to "Hidden tags"#19675](https://github.com/forem/forem/issues/19675) issue, there was a small miscommunication whereby in the mockup we showed that when you "Unhide" a tag it will show as "Following" instead of "Follow".   

## Describe the solution you’d like
However, when we unhide a tag, it should move it to an "unfollowed" state, hence the user should see a "Follow" button. 

## Definition of Done

- [x] When you "Unhide" a tag it will show as "Follow" instead of "Following".
- [x] When you "Unhide" a tag we should unfollow a tag.

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

When we unhide a tag, we should "unfollow" it too, hence the user should see a "Follow" button.

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/20039
- Closes #20039 

## QA Instructions, Screenshots, Recordings

**TAGS PAGE:** 
- Go to /tags and click the "follow" button to follow tags. Refresh the page and ensure that they stay followed.
- Now click on the "Following" button to unfollow the tags. Refresh the page and ensure that they stay unfollowed.
- Now find a tag card with a "Follow" button and a "Hide" button. Click on the "Hide" button to to hide the tag. The "Follow" Button should disappear and an "Unhide" button should show up in place of the "Hide" button.
- Now find a tag card with a "Following" button and a "Hide" button. Click on the "Hide" button to to hide the tag. The "Following" Button should disappear and an "Unhide" button should show up in place of the "Hide" button.
- **_(Main purpose of the PR)_** Now click on the "Unhide" Button to unhide the tags. The "Follow" button will appear and the "Hide" button will appear.

**DASHBOARD PAGES:**

**Dashboard - "Following tags" page:**

- Ensure that you go to /tags and follow many tags. These followed tags will show up on the /dashboard/following_tags. - - - - Click on the "Following" Button to unfollow the tags; the tag should disappear from the page and you should see the nav item count drop by 1.
- Click on the overflow menu button and click on "Hide tag" to hide the tag; the tag should disappear from /dashboard/following_tags but it should be added to /dashboard/hidden_tags and you should see the Following Tags nav item count drop by 1 and the Hidden Tags nav item count increase by 1.

**Dashboard - "Hidden tags" page:**

- **_(Main purpose of the PR)_** Click on the "Unhide" Button to unhide the tags; the tag should disappear from /dashboard/hidden_tags, but it should NOT be added to /dashboard/following_tags and you should NOT see the Following
Tags nav item count increase by 1. The Hidden Tags nav item count decrease by 1. When going back to /tags you should see the tag card for that tag have a "Follow" button.

https://www.loom.com/share/f2a6ba6c47984e00be7651c90817ef53?sid=8d66c79e-adc2-4668-8f1f-7fdc974fd347

### UI accessibility checklist
Functionality change 

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/javascript/tags/Tag.jsx [hunk 1]
```diff
@@ -25,11 +25,19 @@ export const Tag = ({ id, name, isFollowing, isHidden }) => {
...
   const toggleHideButton = () => {
-    setHidden(!hidden);
-    const updatedFollowing = true;
-    setFollowing(updatedFollowing);
+    const updatedHiddenState = !hidden;
+    setHidden(updatedHiddenState);
+
+    // if the tag's new state will be hidden (clicked on the hide button) then we we set it to following.
+    // if the tags new state is to be unhidden (clicked on the unhide button) then we set it to not following.
+    const updatedFollowingState = updatedHiddenState;
```
Comment (Ridhwana) [inline app/javascript/tags/Tag.jsx:33]: This is a bit repetitive but I felt its better to be verbose here to avoid confusion.

**Comment by Ridhwana** [comment]
@anujbhavsar96 even though this is not a design change, I've added you as a reviewer as you picked up the initial issue, feel free to test and leave your feedback as well.

**Review by anujbhavsar96 — APPROVED** [review]
It's working as expected. Thanks!!

**Review by lightalloy — COMMENTED** [review]
I wonder why don't we unfollow on unhiding, but not hiding in the first place?
If a user wants to hide a tag, they don't want to see such posts, does it make sense to keep following it?

**Comment by Ridhwana** [comment]
> I wonder why don't we unfollow on unhiding, but not hiding in the first place?
> 
> If a user wants to hide a tag, they don't want to see such posts, does it make sense to keep following it?

I'm not sure if your question refers to the practical implementation or the theoretical use so I'll try to answer both. 

Practically the way that hiding was architected  is that it's built as 'follow' but instead of positive explicit points, a 'hide' sets negative explicit points. Hence it would be really difficult to 'unfollow' a tag on hide without changing the apps architecture. 

And then theoretically on the feed we show a mixture of tags but when a user 'hides' a tag we will explicitly suppress the posts with tags that are hidden which is different to the behavior of an unfollow.

**Comment by lightalloy** [comment]
> Hence it would be difficult to 'unfollow' a tag on hide without changing the apps architecture and previously recorded data.

Yeah, If we make this change at the moment, there will still be tags that are both hidden and followed at the same time.
But it general, having a tag hidden and followed at the same time (and unfollowing on unhiding as a consequence) seem counterintuitive.

**Comment by Ridhwana** [comment]
> Yeah, If we make this change at the moment, there will still be tags that are both hidden and followed at the same time.
> 
> But it general, having a tag hidden and followed at the same time (and unfollowing on unhiding as a consequence) seem counterintuitive.

Yeah I definitely agree with you, I've had to explain it and document it many times because it's not intuitive concept and would benefit from being changed. 

I think it's something we'd have to go back an architect differently given the time and priority. A small portion of of this counter-intuition is also why we have a complicated front end resulting in feature taking a lot longer than estimated.

**Review by filleduchaos — APPROVED** [review]
🎉 I do think at some point we should revisit tags and all the terminology associated with them, but for now this is about as good as it gets.

**Pull request merged**